src/Umbraco.Cms.ManagementApi/Controllers/NewInstallController.cs

```diff
@@ -0,0 +1,131 @@
+    private readonly IInstallService _installService;
+    private readonly GlobalSettings _globalSettings;
+    private readonly IHostingEnvironment _hostingEnvironment;
+    private readonly InstallHelper _installHelper;
+    private readonly DatabaseBuilder _databaseBuilder;
+
+    public NewInstallController(
+        IUmbracoMapper mapper,
+        IInstallSettingsFactory installSettingsFactory,
+        IInstallService installService,
+        IOptions<GlobalSettings> globalSettings,
+        IHostingEnvironment hostingEnvironment,
+        InstallHelper installHelper,
+        DatabaseBuilder databaseBuilder)
+    {
+        _mapper = mapper;
+        _installSettingsFactory = installSettingsFactory;
+        _installService = installService;
+        _globalSettings = globalSettings.Value;
+        _hostingEnvironment = hostingEnvironment;
+        _installHelper = installHelper;
+        _databaseBuilder = databaseBuilder;
+    }
+
+    [HttpGet("settings")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status428PreconditionRequired)]
+    [ProducesResponseType(typeof(InstallSettingsViewModel), StatusCodes.Status200OK)]
+    public async Task<ActionResult<InstallSettingsViewModel>> Settings()
+    {
+        // Register that the install has started
+        await _installHelper.SetInstallStatusAsync(false, string.Empty);
+
+        InstallSettingsModel installSettings = _installSettingsFactory.GetInstallSettings();
+        InstallSettingsViewModel viewModel = _mapper.Map<InstallSettingsViewModel>(installSettings)!;
+
+        return viewModel;
+    }
+
+    [HttpPost("setup")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status428PreconditionRequired)]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public async Task<IActionResult> Setup(InstallViewModel installData)
+    {
+        if (ModelState.IsValid is false)
+        {
+            return BadRequest(ModelState);
+        }
+
+        InstallData data = _mapper.Map<InstallData>(installData)!;
+        await _installService.Install(data);
+
+        var backOfficePath = _globalSettings.GetBackOfficePath(_hostingEnvironment);
+        return Created(backOfficePath, null);
+    }
+
+    [HttpPost("validateDatabase")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public async Task<IActionResult> ValidateDatabase(DatabaseInstallViewModel viewModel)
+    {
+        if (ModelState.IsValid is false)
+        {
+            return BadRequest(ModelState);
+        }
+
+        DatabaseModel? databaseModel = _mapper.Map<DatabaseModel>(viewModel);
```

Comment:
Since we perform ModelState validation, the mapping cannot fail. Or can it? Consider doing the same as in the `Setup` action, something like `DatabaseModel databaseModel = _mapper.Map<DatabaseModel>(viewModel)!;`

Reply:
Very good point, I even did so in the Setup method above 🤦